keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1159 @@
+
+The ApplySet object should be labeled with:
+
+```yaml
+applyset.k8s.io/id: <applyset-id>
+```
+
+Implicit in this are a few assumptions:
+
+- An object can be part of at most one ApplySet.  This is a limitation, but seems to be a good one in that objects that are part of multiple ApplySets are complicated both conceptually for users and in terms of implementation behaviour.
+- An ApplySet object can be part of another ApplySet (sub-ApplySets).
+
+How the ApplySet object is specified is a tooling decision.  Gitops based tooling may choose to make the ApplySet object explicit in the source git repo.  Other tooling may choose to leverage their existing concepts, for example mapping to a Secret or ConfigMap that they are creating already.  The tooling is responsible for consistently specifying the ApplySet object across apply invocations, so that pruning can be done consistently.
+
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> Do we agree: For kubectl specifically, we propose supporting but not requiring explicitly provided parent objects, with automatic object creation in the latter case. This is explained in more detail below.
+>
+> <<[/UNRESOLVED]>>
+
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Efficient Listing of ApplySets
+
+In order to support listing all the applysets in theory we would need to query all GKs with a label selector.  However, we can reduce the set of GKs that need to be queried with two optimizations:
+
+For built-in types, we limit to ConfigMaps and Secrets.
+For custom resources, we require that CRDs that define types that can be used as ApplySet objects be labeled with a label with a name of `applyset.k8s.io/role/applyset`.
+
+>  <<[UNRESOLVED @justinsb @KnVerey]>>
+>
+> The value is currently unspecified (TODO: maybe v1?)
```

Comment:
@justinsb did we mean that the value might literally be "v1" or that we'd defer having a value initially?

Reply:
I was trying to think that we should do something that is forwards-compatible.  I'm going to send a PR to this PR proposing that people should set the empty value here, so we can define some richer format in future.